# Working On Police data analysis Project Using Python

# Police Dataset

Here,

The data from police check point is given.

The dataset is available on CSV format, We are going to anlayze dataset using Pandas Data Frame.

# Import librabry

In [1]:
import pandas as pd

# Uploading Dataset

In [3]:
df = pd,read_csv("C:\Users\Syed Arif\Downloads\3. Police Data.csv")

SyntaxError: (unicode error) 'unicodeescape' codec can't decode bytes in position 2-3: truncated \UXXXXXXXX escape (2287394368.py, line 1)

# Remove Unicode error

We are Written Small r before Quatation

In [5]:
df = pd.read_csv(r"C:\Users\Syed Arif\Downloads\3. Police Data.csv")

In [6]:
df

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0       1/2/2005      1:55           NaN             M          1985.0   
1      1/18/2005      8:15           NaN             M          1965.0   
2      1/23/2005     23:15           NaN             M          1972.0   
3      2/20/2005     17:15           NaN             M          1986.0   
4      3/14/2005     10:00           NaN             F          1984.0   
...          ...       ...           ...           ...             ...   
65530  12/6/2012     17:54           NaN             F          1987.0   
65531  12/6/2012     22:22           NaN             M          1954.0   
65532  12/6/2012     23:20           NaN             M          1985.0   
65533  12/7/2012      0:23           NaN           NaN             NaN   
65534  12/7/2012      0:30           NaN             F          1985.0   

       driver_age driver_race                   violation_raw  violation  \
0            20.0       White                        Speeding   Speeding   
1            40.0       White                        Speeding   Speeding   
2            33.0       White                        Speeding   Speeding   
3            19.0       White                Call for Service      Other   
4            21.0       White                        Speeding   Speeding   
...           ...         ...                             ...        ...   
65530        25.0       White                        Speeding   Speeding   
65531        58.0       White                        Speeding   Speeding   
65532        27.0       Black  Equipment/Inspection Violation  Equipment   
65533         NaN         NaN                             NaN        NaN   
65534        27.0       White                        Speeding   Speeding   

       search_conducted search_type   stop_outcome is_arrested stop_duration  \
0                 False         NaN       Citation       False      0-15 Min   
1                 False         NaN       Citation       False      0-15 Min   
2                 False         NaN       Citation       False      0-15 Min   
3                 False         NaN  Arrest Driver        True     16-30 Min   
4                 False         NaN       Citation       False      0-15 Min   
...                 ...         ...            ...         ...           ...   
65530             False         NaN       Citation       False      0-15 Min   
65531             False         NaN        Warning       False      0-15 Min   
65532             False         NaN       Citation       False      0-15 Min   
65533             False         NaN            NaN         NaN           NaN   
65534             False         NaN       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65530               False  
65531               False  
65532               False  
65533               False  
65534               False  

[65535 rows x 15 columns]

# Data Preprocessing

# .head()

head is used show to the By default = 5 rows in the dataset

In [7]:
df.head()

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0   1/2/2005      1:55           NaN             M          1985.0   
1  1/18/2005      8:15           NaN             M          1965.0   
2  1/23/2005     23:15           NaN             M          1972.0   
3  2/20/2005     17:15           NaN             M          1986.0   
4  3/14/2005     10:00           NaN             F          1984.0   

   driver_age driver_race     violation_raw violation  search_conducted  \
0        20.0       White          Speeding  Speeding             False   
1        40.0       White          Speeding  Speeding             False   
2        33.0       White          Speeding  Speeding             False   
3        19.0       White  Call for Service     Other             False   
4        21.0       White          Speeding  Speeding             False   

  search_type   stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN       Citation       False      0-15 Min               False  
1         NaN       Citation       False      0-15 Min               False  
2         NaN       Citation       False      0-15 Min               False  
3         NaN  Arrest Driver        True     16-30 Min               False  
4         NaN       Citation       False      0-15 Min               False

# .tail()

tail is used to show  rows by Descending order

In [8]:
df.tail()

stop_date stop_time  country_name driver_gender  driver_age_raw  \
65530  12/6/2012     17:54           NaN             F          1987.0   
65531  12/6/2012     22:22           NaN             M          1954.0   
65532  12/6/2012     23:20           NaN             M          1985.0   
65533  12/7/2012      0:23           NaN           NaN             NaN   
65534  12/7/2012      0:30           NaN             F          1985.0   

       driver_age driver_race                   violation_raw  violation  \
65530        25.0       White                        Speeding   Speeding   
65531        58.0       White                        Speeding   Speeding   
65532        27.0       Black  Equipment/Inspection Violation  Equipment   
65533         NaN         NaN                             NaN        NaN   
65534        27.0       White                        Speeding   Speeding   

       search_conducted search_type stop_outcome is_arrested stop_duration  \
65530             False         NaN     Citation       False      0-15 Min   
65531             False         NaN      Warning       False      0-15 Min   
65532             False         NaN     Citation       False      0-15 Min   
65533             False         NaN          NaN         NaN           NaN   
65534             False         NaN     Citation       False      0-15 Min   

       drugs_related_stop  
65530               False  
65531               False  
65532               False  
65533               False  
65534               False

# .shape

It shoe the total no of rows & Column in the dataset

In [9]:
df.shape

(65535, 15)

# .Columns

It show the no of each Column

In [10]:
df.columns

Index(['stop_date', 'stop_time', 'country_name', 'driver_gender',
       'driver_age_raw', 'driver_age', 'driver_race', 'violation_raw',
       'violation', 'search_conducted', 'search_type', 'stop_outcome',
       'is_arrested', 'stop_duration', 'drugs_related_stop'],
      dtype='object')

 # .dtypes

This Attribute show the data type of each column

In [11]:
df.dtypes

stop_date              object
stop_time              object
country_name          float64
driver_gender          object
driver_age_raw        float64
driver_age            float64
driver_race            object
violation_raw          object
violation              object
search_conducted         bool
search_type            object
stop_outcome           object
is_arrested            object
stop_duration          object
drugs_related_stop       bool
dtype: object

# .unique()

In a column, It show the unique value of specific column.

In [13]:
df['country_name'].unique

<bound method Series.unique of 0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
         ..
65530   NaN
65531   NaN
65532   NaN
65533   NaN
65534   NaN
Name: country_name, Length: 65535, dtype: float64>

In [14]:
df['country_name'].unique()

array([nan])

# nuique()

It will show the total no of unque value from whole data frame.

In [15]:
df.nunique()

stop_date             2651
stop_time             1432
country_name             0
driver_gender            2
driver_age_raw          93
driver_age              73
driver_race              5
violation_raw           12
violation                6
search_conducted         2
search_type             23
stop_outcome             6
is_arrested              2
stop_duration            4
drugs_related_stop       2
dtype: int64

# .describe()

It show the Count, mean , median etc

In [16]:
df.describe()

country_name  driver_age_raw    driver_age
count           0.0    61481.000000  61228.000000
mean            NaN     1967.791106     34.148984
std             NaN      121.050106     12.760710
min             NaN        0.000000     15.000000
25%             NaN     1965.000000     23.000000
50%             NaN     1978.000000     31.000000
75%             NaN     1985.000000     43.000000
max             NaN     8801.000000     88.000000

# .value_counts

It Shows all the unique values with their count

In [17]:
df["driver_race"].value_counts()

White       45747
Black        8267
Hispanic     5611
Asian        1639
Other         211
Name: driver_race, dtype: int64

# Q : 1 ) Remove the column that only contains misssing values

# .isnull()

Find out all the Null values in the data frame

In [19]:
df.isnull().sum()

stop_date                 0
stop_time                 0
country_name          65535
driver_gender          4061
driver_age_raw         4054
driver_age             4307
driver_race            4060
violation_raw          4060
violation              4060
search_conducted          0
search_type           63056
stop_outcome           4060
is_arrested            4060
stop_duration          4060
drugs_related_stop        0
dtype: int64

There are many null values in this data frame
Heighest Null Values is (country_name == 65535 ) Columns
That's Why we Dop the column Country_name

In [20]:
df.drop(columns = country_name , inplace  = True)

NameError: name 'country_name' is not defined

# Why we are facing this error......?

Because Country_name is always written in string form   ==  "country_name"

In [22]:
df.drop(columns = "country_name" , inplace  = True)

In [23]:
df

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
0       1/2/2005      1:55             M          1985.0        20.0   
1      1/18/2005      8:15             M          1965.0        40.0   
2      1/23/2005     23:15             M          1972.0        33.0   
3      2/20/2005     17:15             M          1986.0        19.0   
4      3/14/2005     10:00             F          1984.0        21.0   
...          ...       ...           ...             ...         ...   
65530  12/6/2012     17:54             F          1987.0        25.0   
65531  12/6/2012     22:22             M          1954.0        58.0   
65532  12/6/2012     23:20             M          1985.0        27.0   
65533  12/7/2012      0:23           NaN             NaN         NaN   
65534  12/7/2012      0:30             F          1985.0        27.0   

      driver_race                   violation_raw  violation  \
0           White                        Speeding   Speeding   
1           White                        Speeding   Speeding   
2           White                        Speeding   Speeding   
3           White                Call for Service      Other   
4           White                        Speeding   Speeding   
...           ...                             ...        ...   
65530       White                        Speeding   Speeding   
65531       White                        Speeding   Speeding   
65532       Black  Equipment/Inspection Violation  Equipment   
65533         NaN                             NaN        NaN   
65534       White                        Speeding   Speeding   

       search_conducted search_type   stop_outcome is_arrested stop_duration  \
0                 False         NaN       Citation       False      0-15 Min   
1                 False         NaN       Citation       False      0-15 Min   
2                 False         NaN       Citation       False      0-15 Min   
3                 False         NaN  Arrest Driver        True     16-30 Min   
4                 False         NaN       Citation       False      0-15 Min   
...                 ...         ...            ...         ...           ...   
65530             False         NaN       Citation       False      0-15 Min   
65531             False         NaN        Warning       False      0-15 Min   
65532             False         NaN       Citation       False      0-15 Min   
65533             False         NaN            NaN         NaN           NaN   
65534             False         NaN       Citation       False      0-15 Min   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65530               False  
65531               False  
65532               False  
65533               False  
65534               False  

[65535 rows x 14 columns]

# Q : 2 ) For Spreeding , were men or women stopped more often ?

In [29]:
df[df.violation == "Speeding"].driver_gender.value_counts()

M    25517
F    11686
Name: driver_gender, dtype: int64

# Q : 3 ) Who gender affect who gets searched during a stop ?

In [31]:
df.groupby("driver_gender").search_conducted.sum()

driver_gender
F     366
M    2113
Name: search_conducted, dtype: int64

# Q : 4 ) What is mean stop_duration ?

In [36]:
df['stop_duration'].mean()

TypeError: can only concatenate str (not "int") to str

# Reason of this type of error ?

Beacause the column Stop_duration is String format , Before we take an output it will convert string to int

In [37]:
df.stop_duration.value_counts()

0-15 Min     47379
16-30 Min    11448
30+ Min       2647
2                1
Name: stop_duration, dtype: int64

In [39]:
df['stop_duration'] = df["stop_duration"].map({"0-15 Min" : 8 , "16-30 Min" : 25 , "30+ Min " : 40} )

In [40]:
df

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
0       1/2/2005      1:55             M          1985.0        20.0   
1      1/18/2005      8:15             M          1965.0        40.0   
2      1/23/2005     23:15             M          1972.0        33.0   
3      2/20/2005     17:15             M          1986.0        19.0   
4      3/14/2005     10:00             F          1984.0        21.0   
...          ...       ...           ...             ...         ...   
65530  12/6/2012     17:54             F          1987.0        25.0   
65531  12/6/2012     22:22             M          1954.0        58.0   
65532  12/6/2012     23:20             M          1985.0        27.0   
65533  12/7/2012      0:23           NaN             NaN         NaN   
65534  12/7/2012      0:30             F          1985.0        27.0   

      driver_race                   violation_raw  violation  \
0           White                        Speeding   Speeding   
1           White                        Speeding   Speeding   
2           White                        Speeding   Speeding   
3           White                Call for Service      Other   
4           White                        Speeding   Speeding   
...           ...                             ...        ...   
65530       White                        Speeding   Speeding   
65531       White                        Speeding   Speeding   
65532       Black  Equipment/Inspection Violation  Equipment   
65533         NaN                             NaN        NaN   
65534       White                        Speeding   Speeding   

       search_conducted search_type   stop_outcome is_arrested  stop_duration  \
0                 False         NaN       Citation       False            8.0   
1                 False         NaN       Citation       False            8.0   
2                 False         NaN       Citation       False            8.0   
3                 False         NaN  Arrest Driver        True           25.0   
4                 False         NaN       Citation       False            8.0   
...                 ...         ...            ...         ...            ...   
65530             False         NaN       Citation       False            8.0   
65531             False         NaN        Warning       False            8.0   
65532             False         NaN       Citation       False            8.0   
65533             False         NaN            NaN         NaN            NaN   
65534             False         NaN       Citation       False            8.0   

       drugs_related_stop  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  
...                   ...  
65530               False  
65531               False  
65532               False  
65533               False  
65534               False  

[65535 rows x 14 columns]

In [41]:
df["stop_duration"].mean()

11.308276811668112

# Compare the age distribuation of each violation ?

In [44]:
df.groupby("violation").driver_age.describe()

count       mean        std   min   25%   50%   75%  \
violation                                                                    
Equipment             6507.0  31.682957  11.380671  16.0  23.0  28.0  39.0   
Moving violation     11876.0  36.736443  13.258350  15.0  25.0  35.0  47.0   
Other                 3477.0  40.362381  12.754423  16.0  30.0  41.0  50.0   
Registration/plates   2240.0  32.656696  11.150780  16.0  24.0  30.0  40.0   
Seat belt                3.0  30.333333  10.214369  23.0  24.5  26.0  34.0   
Speeding             37120.0  33.262581  12.615781  15.0  23.0  30.0  42.0   

                      max  
violation                  
Equipment            81.0  
Moving violation     86.0  
Other                86.0  
Registration/plates  74.0  
Seat belt            42.0  
Speeding             88.0